src/PJ_deformation.c

```diff
@@ -0,0 +1,294 @@
+corrections are then applied to the input coordinates in cartesian
+space.
+
+A full deformation model is described by two grids, one for the horizontal
+components and one for the vertical component. The horizontal grid is
+stored in CTable/CTable2 and the vertical grid is stored in the GTX
+format. Both grids are expected to contain grid-values in units of
+m/year.
+
+************************************************************************
+* Copyright (c) 2017, Kristian Evers
+*
+* Permission is hereby granted, free of charge, to any person obtaining a
+* copy of this software and associated documentation files (the "Software"),
+* to deal in the Software without restriction, including without limitation
+* the rights to use, copy, modify, merge, publish, distribute, sublicense,
+* and/or sell copies of the Software, and to permit persons to whom the
+* Software is furnished to do so, subject to the following conditions:
+*
+* The above copyright notice and this permission notice shall be included
+* in all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+* OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+* THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+* DEALINGS IN THE SOFTWARE.
+*
+***********************************************************************/
+#define PJ_LIB__
+#include <errno.h>
+#include <proj.h>
+#include "proj_internal.h"
+#include "projects.h"
+
+PROJ_HEAD(deformation, "Kinematic grid shift");
+
+#define TOL 1e-8
+#define MAX_ITERATIONS 10
+
+struct pj_opaque {
+    double t_obs;
+    double t_epoch;
+    int has_xy_grids, has_z_grids;
+    PJ *cart;
+};
+
+static XYZ get_xygrid_shift(PJ* P, XYZ cartesian) {
+    PJ_COORD geodetic, shift;
+
+    geodetic.lpz = pj_inv3d(cartesian, P->opaque->cart);
+    shift.lp = proj_hgrid_value(P, geodetic.lp);
+
+    return shift.xyz;
+}
+
+static double get_zgrid_shift(PJ* P, XYZ cartesian) {
+    PJ_COORD geodetic;
+
+    geodetic.lpz = pj_inv3d(cartesian, P->opaque->cart);
+
+    return proj_vgrid_value(P, geodetic.lp);
+}
+
+static XYZ reverse_hshift(PJ *P, XYZ input, double dt) {
+    XYZ out, delta, dif;
+    int i = MAX_ITERATIONS;
+
+    delta = get_xygrid_shift(P, input);
+
+    out.x = input.x + dt*delta.x;
+    out.y = input.y - dt*delta.y;
+    out.z = input.z;
+
+    do {
+        delta = get_xygrid_shift(P, out);
+
+        if (delta.x == HUGE_VAL)
+            break;
+
+        dif.x  = out.x - dt*delta.x - input.x;
+        dif.y  = out.y + dt*delta.y - input.y;
+        out.x += dif.x;
+        out.y += dif.y;
+
+    } while ( i-- && fabs(dif.x) > TOL && fabs(dif.y) > TOL );
```

Comment:
If I understand the algorithm correctly, the dif.x,y cases should be or'ed, rather than and'ed: Otherwise you can exit the loop as soon you reach the right coordinate in one axis, while still being kilometers off along the other.

Addititonally, it might be slightly better to use hypot (dif.x, dif.y) > TOL (and perhaps adjust TOL accordingly), rather than the dual fabs: Since we can have no idea about the local gradients, we really want to make sure we're "inside the disk", while checking each dimension individually only ensures we're "inside the square".

Finally, I think `i--` should be changed to `--i`, otherwise `i = MAX_ITERATIONS` would result in `MAX_ITERATIONS + 1` iterations.

Reply:
This was taken more or less verbatim from the implementation in `nad_cvt()`. I've changed the variable names, but otherwise it is the same algorithm. I think your suggestions are sensible and seem to be working fine as well. I don't think the decrement method of `i` matters all that much since convergence is reached within the first few passes of the loop but we might as well do it as correct as possible.